src/cmd/serve.rs

```diff
@@ -763,55 +810,153 @@ pub fn serve(
         };
     }
 }
 
+/// Filter `notify_debouncer_full` events. For events that we care about,
+/// return our internal simplified representation. For events we don't care about,
+/// return `None`.
+fn get_relevant_event_kind(event_kind: &EventKind) -> Option<SimpleFSEventKind> {
+    match event_kind {
+        EventKind::Create(create_kind) => match create_kind {
+            CreateKind::File | CreateKind::Folder => Some(SimpleFSEventKind::Create),
+            _ => None,
+        },
+        EventKind::Modify(modify_kind) => {
+            match modify_kind {
+                ModifyKind::Data(data_change) => match data_change {
+                    DataChange::Size | DataChange::Content => Some(SimpleFSEventKind::Modify),
+                    _ => None,
+                },
+                ModifyKind::Metadata(metadata_kind) => {
+                    // IntelliJ can modify file metadata; account for that here.
+                    // https://github.com/passcod/notify/issues/150#issuecomment-494912080
+                    match metadata_kind {
+                        MetadataKind::WriteTime
+                        | MetadataKind::Permissions
+                        | MetadataKind::Ownership => Some(SimpleFSEventKind::Modify),
+                        _ => None,
+                    }
+                }
+                ModifyKind::Name(rename_mode) => match rename_mode {
+                    RenameMode::To => Some(SimpleFSEventKind::Modify),
+                    _ => None,
+                },
+                _ => None,
+            }
+        }
+        EventKind::Remove(remove_kind) => match remove_kind {
+            RemoveKind::File | RemoveKind::Folder => Some(SimpleFSEventKind::Remove),
+            _ => None,
+        },
+        _ => None,
+    }
+}
+
+/// Return a new vector with no duplicate `PathBuf` elements. For any duplicate, keep the `PathBuf`
+/// with the latest `Instant` value.
+fn debounce_events(
+    events: &Vec<(PathBuf, SimpleFSEventKind, Instant)>,
+) -> Vec<(PathBuf, SimpleFSEventKind)> {
+    let mut filtered = events.to_vec();
+
+    // Could use `{sort|dedup}_by_key` but then would need an annotation function for lifetimes.
+    // https://users.rust-lang.org/t/lifetimes-with-vec-dedup-by-key/65439
+    filtered.sort_by(|(p1, _, t1), (p2, _, t2)| {
+        if p1 == p2 {
+            // Compare t2 as first operand for descending timestamp sort.
+            if t2 > t1 {
+                Ordering::Greater
+            } else if t2 < t1 {
+                Ordering::Less
+            } else {
+                Ordering::Equal
+            }
+        } else {
+            p1.cmp(p2)
+        }
+    });
+    filtered.dedup_by(|(p1, _, _), (p2, _, _)| p1 == p2);
+
+    filtered.into_iter().map(|(p, e, _)| (p, e)).collect()
+}
+
+/// Return a map storing all events of a like `ChangeKind` behind one key.
+/// Each element maps `change_kind -> (partial_path, full_path, event_kind)`.
+fn map_events_to_change_map<'a>(
```

Comment:
let's move all that code to another file

Reply:
Let's put it in the root component, next to `cli.rs`

Reply:
Done.